### PR TITLE
FIX Validate Category / Tag Titles 

### LINKED ISSUE #695: The user can create and save a Blog Category or Blog Tag without a Title
### Description
When a user creates a blog, in the Categorisation tab he can add a Blog Category or Blog Tag without a title and then save empty values.

### AC's
- Validation that the Title field in Tag / Category is valid at the time the blog post is saved is added;
- Tests checking that these fields are not empty are added;

### PR DESCRIPTION
**Pull request by freezernick**
Closes #695

### CODE REVIEW

File: src/Model/BlogObject.php [hunk 1]
```diff
@@ -72,6 +72,10 @@ public function validate()
             $validation->addError($this->getDuplicateError(), self::DUPLICATE_EXCEPTION);
         }
 
+        if(empty($this->Title)) {
```
Comment (michalkleiner) [inline src/Model/BlogObject.php]:
```suggestion
        if (empty($this->Title)) {
```

**Review by GuySartorelli — COMMENTED** [review]
@silverstripe/core-team 
What's our thoughts about this in regards to semver?
Specifically, changing the validation rules for whether an empty value is valid or not.

On one hand, it seems like there's no valid reason to have an empty category/tag - but on the other, someone could very well be going `BlogTag::create()->write()` in a unit test, if the _value_ of the tag doesn't matter for that test.

File: src/Model/BlogObject.php [hunk 2]
```diff
@@ -72,6 +72,10 @@ public function validate()
             $validation->addError($this->getDuplicateError(), self::DUPLICATE_EXCEPTION);
         }
 
+        if (empty($this->Title)) {
+            $validation->addError($this->getEmptyTitleError(), self::EMPTY_TITLE_EXCEPTION);
```
Comment (ChloeSartorelli) [inline src/Model/BlogObject.php:76]:
```suggestion
            $validation->addError($this->getEmptyTitleError(), self::EMPTY_TITLE_EXCEPTION ?? 'EMPTY_TITLE');
```
The constant won't be declared on any custom classes that use this trait, and requiring it to be added is technically a breaking change.

Note: Please test that this actually works - I've not tried the null coalescing operator in this context before. I think it'll work but it'll pay to check.

**Review by GuySartorelli — CHANGES_REQUESTED** [review]

File: src/Model/BlogObject.php (hunk 2 above)
Comment (emteknetnz) [reply src/Model/BlogObject.php:76]: Yikes, shouldn't be calling `self::<const>` on a trait.  Not a great choice when this was made back in the day

I don't think null coalescing works here

```php
class C { function F(){ var_dump(self::Z ?? 'default'); } }; (new C)->F();
// PHP Warning:  Uncaught Error: Undefined constant C::Z in php shell cod
```

**Comment by emteknetnz** [comment]
>> someone could very well be going BlogTag::create()->write() in a unit test

No one other than core team is going to write new unit tests, and the existing unit tests are green on this PR, so I see no reason to be concerned here.

**Comment by GuySartorelli** [comment]
> No one other than core team is going to write new unit tests

People write unit tests for their own projects and modules all the time. It's not super likely those tests will rely on blog tags or categories but it's also not impossible.

**Comment by emteknetnz** [comment]
Oh sorry I meant "No one other than core team is going to write new unit tests _for the blog module_" :-) 

What's the concern? That a project will create their own unit test that instantiates a their own subclass of a  `BlogCategory` or `BlogTag` with a blank Title?  That's seems extremely unlikely.  And even if they did and their test fails, they'll get a pretty clear `Title must not be empty` error which should be easy to trace back, so fixing the broken tests should require minimal effort.

While we're very strict with respecting semver API in minor releases, historically we've been somewhat lenient/pragmatic with behavioral changes, including changes to default configuration values.  This PR seems at the low end of changes within a minor.

**Comment by michalkleiner** [comment]
> What's our thoughts about this in regards to semver? Specifically, changing the validation rules for whether an empty value is valid or not.
> 
> On one hand, it seems like there's no valid reason to have an empty category/tag - but on the other, someone could very well be going `BlogTag::create()->write()` in a unit test, if the _value_ of the tag doesn't matter for that test.

Not bothered by this particular change, I agree with Steve on this.

**Comment by GuySartorelli** [comment]
Sweet, just wanted to get the general feeling about it, since I was on the fence about it.
There's problems with how the constant is being applied, but sounds like the general approach is okay.

**Comment by GuySartorelli** [comment]
@freezernick Are you still interested in this pull request being merged? If so, please make changes to how the constant is applied as requested above.

**Comment by GuySartorelli** [comment]
I'm closing this due to inactivity. If you want to work on it again I'll be happy to reopen it, or you can open a new pull request from scratch.

**Pull request closed without merging**